### PR TITLE
Add MeanVarNormalization for use by BatchNormalization for training

### PR DESCRIPTION
**Pull request by jfix71**
I removed the CompilationMode from `lower()` in a separate commit here as well. This resolves https://github.com/pytorch/glow/issues/1309.

### CODE REVIEW

File: tools/ClassGen/NodeGen.cpp [hunk 1]
```diff
@@ -120,6 +120,17 @@ int main(int argc, char **argv) {
                     "Momentum. Similar to Caffe2 SpatialBN, and ONNX "
                     "BatchNormalization operator.");
 
+  BB.newNode("MeanVarNormalization")
+      .addInput("Input")
+      .addInput("Mean")
+      .addInput("Var")
+      .addMember(MemberType::SizeT, "ChannelIdx")
+      .addMember(MemberType::Float, "Momentum")
+      .addResultFromCtorArg("NewMean")
```
Comment (artemrakhov) [inline tools/ClassGen/NodeGen.cpp]: I think you can write here:
```
       .addResult("Mean.getType()")
       .addResult("Var.getType()")
```
Then you won't have to pass types to the constructor.

Comment (jfix71) [reply tools/ClassGen/NodeGen.cpp]: I don't think this is currently supported for Nodes with multiple outputs. They're both called "Result" and so `getResult()` is defined twice. Could fix this in a future PR.

Comment (qcolombet) [reply tools/ClassGen/NodeGen.cpp]: You’ll need to give them a name:
.addResult(“type”, “name”)

Comment (jfix71) [reply tools/ClassGen/NodeGen.cpp]: I didn't realize we supported that! Done.

**Comment by jfix71** [comment]
I realized that the save nodes should have been added during differentiation and not lowering, otherwise the variables would incorrectly be seen as not written into post-differentiation/pre-lowering.

File: lib/Graph/Grad.cpp [hunk 2]
```diff
@@ -190,6 +189,43 @@ Function *glow::differentiate(Function *F, const TrainingConfig &conf,
...
+      // variables. These will be used during the next iteration of training.
+      G->createSave("saveMean", MVN->getNewMean(),
+                    llvm::cast<Variable>(mean.getNode()));
+      G->createSave("saveVar", MVN->getNewVar(),
+                    llvm::cast<Variable>(var.getNode()));
+
+      // Replace the BN's mean and variance with the new mean and variance
+      // calculated from MVN.
+      for (size_t i = 0; i < BN->getNumInputs(); i++) {
+        if (BN->getNthInput(i) == mean) {
```
Comment (qcolombet) [inline lib/Graph/Grad.cpp]: Could we directly modify the operands we believe are mean and var while asserting on them being equal instead of using a loop.
I.e.,
assert(BN->getNthInput(idxMean) ==  mean);
BN->setNthInput(idxMean, mean);

Comment (jfix71) [reply lib/Graph/Grad.cpp]: Yeah, I could make that change if desired. Seems a little fragile as its based on the order of operands in NodeGen but the assert should catch any issues.

Comment (jfix71) [reply lib/Graph/Grad.cpp]: Done.

**Comment by nadavrot** [comment]
@jfix71 I notices that we don't have any neural networks that use batchednormalization, except for the grad checks. I think it would be a good idea to add at least one compile test that checks that things work as expected across the compilation lifetime (creation->differentiation->execution)
